### PR TITLE
upgrade jni v0.21

### LINKED ISSUE #22: jni 0.21 migration
Presently this crate uses [`jni-rs`](https://github.com/jni-rs/jni-rs) v0.19 as an optional dependency for doc generation, and a required dependency for the Android OS target.

The latest release stream, 0.21.x, "makes extensive breaking changes in order to improve safety. Most projects that use this library will need to be changed accordingly.". The upstream project offers a [migration guide](https://github.com/jni-rs/jni-rs/blob/master/docs/0.21-MIGRATION.md) to help with this process.

Given the advertised safety improvements it's likely worthwhile to consider the steps required for migration.

### PR DESCRIPTION
**Pull request by flisky**
~jni v0.19 doesn't work for me.~

~I use tokio multi threads runtime, and our app is killed by system -~

```
e.android.debug: java_vm_ext.cc:594] JNI DETECTED ERROR IN APPLICATION: a thread (tid 5917 is making JNI calls without being attached
e.android.debug: java_vm_ext.cc:594]     in call to PushLocalFrame
```
~It turns out `java_vm::get_env()` goes null even after calling `AttachCurrentThread` in a few threads, while other threads are okay.~
~I have no idea. However, upgrade to jni v0.21.0 eliminates the problem for me.~

fixes #22

### CODE REVIEW

**Comment by complexspaces** [comment]
Thanks for the PR! It looks like the changes here aren't [currently compiling for Android](https://github.com/rustls/rustls-platform-verifier/actions/runs/7541107318/job/20530607624?pr=63). Once CI is passing I'm happy to take a look through this.

> It turns out java_vm::get_env() goes null even after calling AttachCurrentThread in a few threads, while other threads are okay.
> I have no idea. However, upgrade to jni v0.21.0 eliminates the problem for me.

That's very strange, and definitely something I'll look into on the side. We use this exact same JNI-abstraction inside 1Password's Android app (this code was copy/pasted out) and we've never seen this issue AFAICT. If you have any other JNI code in your app, it might be worth checking if its having a poor interaction with `rustls-platform-verifier`.

**Comment by flisky** [comment]
> If you have any other JNI code in your app, it might be worth checking if its having a poor interaction with rustls-platform-verifier.

We're using `mozilla/uniffi` which uses JNA actually. We have to configure JNI environment mannually, and https://github.com/mozilla/uniffi-rs/issues/1778#issuecomment-1807979746 inspires us.

File: rustls-platform-verifier/src/android.rs [hunk 1]
```diff
@@ -165,17 +166,16 @@ impl<'a> Context<'a> {
...
 
     // 16 is the default capacity in the JVM, we can make this configurable if necessary
-    env.push_local_frame(16)?;
+    context.env.push_local_frame(16)?;
 
-    let res = f(&context);
+    let res = f(&mut context);
 
-    env.pop_local_frame(JObject::null())?;
+    unsafe { context.env.pop_local_frame(&JObject::null()) }?;
```
Comment (complexspaces) [inline rustls-platform-verifier/src/android.rs:178]: This should have a `// SAFETY:` comment attached to it now, but I'm attempting to determine the soundness requirements. [The documentation](https://docs.rs/jni/latest/jni/struct.JNIEnv.html#method.pop_local_frame) says local references become invalid, and I assume those are JVM references and not anything related to native code.

Given that we only return native Rust objects from `f`, I think we follow that rule correctly. If you happen to know, does that sound reasonable to you as well?

**Review by complexspaces — CHANGES_REQUESTED** [review]

**Comment by complexspaces** [comment]
> It turns out java_vm::get_env() goes null even after calling AttachCurrentThread in a few threads, while other threads are okay.
> I have no idea. However, upgrade to jni v0.21.0 eliminates the problem for me.

We also use a multi-threaded Tokio runtime at work and haven't encountered this, so I am curious where the difference is.

@flisky Are you able to/do you mind sharing your Android context initialization code? While looking at the SO post in your linked GitHub comment, some things stood out to me as strange. Notably its unclear why they are wrangling the list of JVMs and current thread instead of passing an environment from Kotlin/Java and using `JniEnv::get_java_vm` (which calls `GetJavaVM`) instead. You can see an example of this [here](https://github.com/rustls/rustls-platform-verifier/blob/main/rustls-platform-verifier/src/tests/ffi.rs#L87-L100) for this crate's tests.

**Comment by flisky** [comment]
I create a demo project for this, and you can take a look at https://github.com/flisky/jnidemo/blob/main/src/android.rs
Unfortunately, I cannot reproduce with that project. And I'll tweak it in next few days.

File: rustls-platform-verifier/src/android.rs [hunk 2]
```diff
@@ -126,34 +126,35 @@ impl From<JNIError> for Error {
...
+    pub(super) fn env(&mut self) -> &mut JNIEnv<'a> {
+        &mut self.env
     }
 
     /// Borrow the `applicationContext` from the Android application
     /// <https://developer.android.com/reference/android/app/Application>
-    pub(super) fn application_context(&self) -> &JObject<'a> {
-        &self.context
+    pub(super) fn application_context(&self) -> GlobalRef {
+        self.context.clone()
```
[the remark this answers is not in the document]
Comment (flisky) [reply rustls-platform-verifier/src/android.rs:142]: I'm afraid not. The `context.env` is a mutable reference, so other context fields must be full owned.

File: rustls-platform-verifier/src/android.rs [hunk 3]
```diff
@@ -126,34 +126,35 @@ impl From<JNIError> for Error {
...
     }
 
     /// Load a class from the application class loader
     ///
     /// This should be used instead of `JNIEnv::find_class` to ensure all classes
     /// in the application can be found.
-    pub(super) fn load_class(&self, name: &str) -> Result<JClass<'a>, Error> {
+    pub(super) fn load_class(&mut self, name: &str) -> Result<JClass<'a>, Error> {
+        let loader = self.loader.clone();
```
[the remark this answers is not in the document]
Comment (flisky) [reply rustls-platform-verifier/src/android.rs:150]: I think it's the same problem as the above question.

File: rustls-platform-verifier/src/android.rs (hunk 1 above)
Comment (flisky) [reply rustls-platform-verifier/src/android.rs:178]: Yes, I agree. Maybe it's better to declare `T: 'static` to make sure it's native rust object?

**Comment by flisky** [comment]
Hey @complexspaces, I'm back on this issue, and I think I might find the root case.

All working threads are initiated from java code, and `GetEnv` goes `Ok` without calling `AttachCurrentThread`,
while the panicked thread is initiated from rust code, and `AttachCurrentThread` doesn't `Err`, but calling `PushLocalFrame` panic.

```
05-22 22:34:04.098 23247 23884 D jni::wrapper::java_vm::vm: message=calling unchecked JavaVM method: GetEnv
05-22 22:34:04.098 23247 23884 D jni::wrapper::java_vm::vm: message=looking up JavaVM method GetEnv
05-22 22:34:04.099 23247 23884 D jni::wrapper::java_vm::vm: message=found JavaVM method
05-22 22:34:04.099 23247 23884 D jni::wrapper::java_vm::vm: message=calling unchecked JavaVM method: AttachCurrentThread
05-22 22:34:04.100 23247 23884 D jni::wrapper::java_vm::vm: message=looking up JavaVM method AttachCurrentThread
05-22 22:34:04.100 23247 23884 D jni::wrapper::java_vm::vm: message=found JavaVM method
05-22 22:34:04.100 23247 23884 D jni::wrapper::java_vm::vm: message=Attached thread async-compat/tokio-1 (ThreadId(1)). 1 threads attached
05-22 22:34:04.101 23247 23884 D jni::wrapper::jnienv: message=calling unchecked jni method: PushLocalFrame
05-22 22:34:04.102 23247 23884 D jni::wrapper::jnienv: message=looking up jni method PushLocalFrame
05-22 22:34:04.102 23247 23884 D jni::wrapper::jnienv: message=found jni method
05-22 22:34:04.102 23247 23884 F dge.app.android: java_vm_ext.cc:594] JNI DETECTED ERROR IN APPLICATION: a thread (tid 23884 is making JNI calls without being attached
05-22 22:34:04.102 23247 23884 F dge.app.android: java_vm_ext.cc:594]     in call to PushLocalFrame
```

Google gives me this: https://groups.google.com/g/android-ndk/c/9Ao-yEE-i3c/m/CC6AW_MuT1MJ

Maybe I need find some wworkaround way, and it would be very appreciated if you have any thought on this.

**Comment by cpu** [comment]
I'm not sure I understand the current state of this PR? Are changes required? Is the update no longer a priority since [a workaround](https://github.com/rustls/rustls-platform-verifier/pull/63#issuecomment-2125127857) was discovered?

**Comment by flisky** [comment]
I think it's waiting for upstreaming dependency coordination: jni-rs/jni-rs#513, and yes, it's not a priority.  
We could revisit it when upsteam goes ready.

**Pull request closed without merging**